src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,426 @@
+                    end = 0;
+                    Debug.Fail("Complete statement not supported on For statement iterator");
+                }
+            }
+            else
+            {
+                end = currentNode.Span.End;
+            }
+
+            // If a node's semicolon is missing, the trailing trivia (including new line) is associated with the last token.
+            // To avoid placing the semicolon on the next line or after comments, we need to adjust the position.
+            return root.FindToken(end).GetPreviousToken().Span.End;
+        }
+
+        private static bool CaretIsInForStatementDeclaration(int caretPosition, ForStatementSyntax forStatementSyntax)
+        {
+            return caretPosition > forStatementSyntax.Declaration.Span.Start && caretPosition < forStatementSyntax.Declaration.Span.End;
+        }
+
+        private static bool CaretIsInForStatementCondition(int caretPosition, ForStatementSyntax forStatementSyntax)
+        {
+            return caretPosition > forStatementSyntax.Condition.SpanStart && caretPosition < forStatementSyntax.Condition.Span.End;
+        }
+
+        /// <summary>
+        /// Examines the enclosing statement-like syntax for an expression which is eligible for statement completion.
+        /// </summary>
+        /// <remarks>
+        /// <para>This method tries to identify <paramref name="currentNode"/> as a node located within an argument
+        /// list, where the immediately-containing statement resembles an "expression statement". This method returns
+        /// <see langword="true"/> if the node matches a recognizable pattern of this form.</para>
+        /// </remarks>
+        private static bool LooksLikeNodeInArgumentList(SyntaxNode currentNode,
+            int caret, bool isCaretAtEndOfLine, ISyntaxFactsService syntaxFacts)
+        {
+            // work our way up the tree, looking for a node of interest within the current statement
+            bool nodeFound = false;
+            while (!IsStatementOrFieldDeclaration(currentNode, syntaxFacts))
+            {
+                // Look for the following types of nodes:
+                // ArgumentList: covers method invocations like object.Method(arg)
+                // ArrayRankSpecifier: covers new Type[dim]
+                // ElementAccessExpression: covers indexer invocations like array[index]
+                // ParenthesizedExpression: covers (3*(x+y))
+                if (currentNode.IsKind(SyntaxKind.ArgumentList, SyntaxKind.ArrayRankSpecifier, SyntaxKind.ElementAccessExpression, SyntaxKind.ParenthesizedExpression))
+                {
+                    nodeFound = true;
+                }
+
+                // No special action is performed at this time if `;` is typed inside a string, including
+                // interpolated strings.  
+                if (IsInAString(currentNode, isCaretAtEndOfLine))
```

Comment:
I do not know why the conversation was resolved.  The primary complaint still stands.  We have IsInNonUserCode.  We should use that instead of having specialzied logic just for the string case.

Or, if this code is more appropriate, it needs a comment explaining why it had to have specialized logic here.

Reply:
`IsInNonUserCode` returns false inside of interpolated strings, so we can't get rid of the current check. We can _add_ a check for `IsInNonUserCode` if a test is added at the same time and that test would fail without the added check.

Reply:
I'm referring to this: `$"{obj.ToString($$)}"`